### PR TITLE
Traffic Manager version 2017-05-01

### PR DESCRIPTION
**Pull request by vladca**
Note: review commit-by-commit to get a good diff

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

---
This change is [Reviewable](https://reviewable.io/reviews/azure/azure-rest-api-specs/1180)

### CODE REVIEW

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (alvadb) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Noticed that there aren't any required properties.  Is this really the case?

**Review by alvadb — CHANGES_REQUESTED** [review]

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vladca) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Yes indeed, that's the case for the default geo hierarchy

**Review by alvadb — COMMENTED** [review]
The linter and model job have multiple failures.  Please look at the errors and resolve.

**Review by alvadb — APPROVED** [review]

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: I'd like to point our few things here:

1. Are you aware that changing the operation names is a breaking change in SDKs
2. Is this a expected change?
3. One of the SDK that would break is python [azure-mgmt-trafficmanager](https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt-trafficmanager/azure/mgmt/trafficmanager/traffic_manager_management_client.py#L88)

**Review by vishrutshah — CHANGES_REQUESTED** [review]
@vladca Could you please confirm why are you making the breaking change? I'd be happy to merge the PR once i get more insights. Thanks!

**Comment by vladca** [comment]
Hi Vishrut, this change was wrongly introduced in 2017-03-01. It was a mistake. I am reverting it in 2017-03-01 and in the newly added 2017-05-01.

**Comment by vladca** [comment]
Hi Vishrut, this change was wrongly introduced in 2017-03-01. It was a mistake. I am reverting it in 2017-03-01 and in the newly added 2017-05-01.

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: There should not be any `Gets` / `Sets` in the description. It's highly discouraged because that is very C# specific and when we generate documentation from here, customer doesn't like them. Let's make this better reference docs.

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: `id`, `name` & `type` should be `readOnly: true`

Modeling Reference: https://github.com/Azure/azure-rest-api-specs-pr/blob/master/common-types/resource-management/v1/types.json#L3
RPCViolation: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/227979765#L1123

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Please model `TrackedResource` as https://github.com/Azure/azure-rest-api-specs-pr/blob/master/common-types/resource-management/v1/types.json#L23 to avoid duplication and centralize model properties

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Please add `description` element onto `SubResource` model

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Please add `description` to `CloudErrorBody` model

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Just a warning: Booleans are not descriptive and make them hard to use. Instead use string enums with allowed set of values defined.

RPCViolation: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/227979765#L1173

Did you get approval from API Reviewer Board?

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Why doesn't we `Security` & `securityDefinitions` for this REST api?

Reference: https://github.com/Azure/azure-rest-api-specs/blob/master/arm-network/2017-03-01/swagger/applicationGateway.json#L18

SDKViolations: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/227979765#L1290

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: There are many `description` elements missing. Please refer the CI to fix and help us make the great documentation from this API :)

Example: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/227979765#L1499 Search for word `DescriptionMissing `

**Review by vishrutshah — CHANGES_REQUESTED** [review]
@vladca I've left few important comments to improve the swagger as we are moving with new api-version that is our best opportunity to improve the SDK + CLI + Documentation experience. Let us know what you think. Thanks!

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vladca) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: We are a different RP within the same RP namespace. Different team, different codebase, different service than the rest of Microsoft.Network; the only thing we share is the namespace and ARM manifest

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vladca) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: It's not a miss, that is a tenant only resource, same as the geographic hierarchy

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vladca) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Endpoint is not a tracked ARM resource, it is a non-tracked child resource, e.g. it does not have a location, tags etc.

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Hmm Interesting. Good to know :)

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: I am bit confused here. If you are defining a parameter referenced as `SubscriptionIdParameter ` in `parameters` section which is by [definition](https://github.com/Azure/azure-rest-api-specs/pull/1180/files/b1f877954e737e88c943c184dd41a394b007ab29#diff-09eb574a88e842533ca4d6116eb9516aR957) is `"in": "path"` & `"required": true` and REST path that `/providers/Microsoft.Network/checkTrafficManagerNameAvailability` does not have that parameters at all. How are we going to take that as input? 

Could you please help me by providing example curl / fiddler example targeting this path and also taking `SubscriptionIdParameter ` as part of it, to better understand it?

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: I agree. That is why we should model `Resource` as `id`, `name` & `type` only and use it here. After that we should have another model named as `TrackedResource` inheriting from `Resource` and add `location` & `tags` as property and use while modeling all the tracked resource.

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Where `SubResource` is this getting used?

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Should this be modeled as `Enum` ?

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Should this be modeled as `Enum`?

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Is this supposed to be `readOnly: true` as the description suggests `This is formed from the...`

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: minor: typo `.. `

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Should be modeled with `minimum` & `maximum` like https://github.com/Azure/azure-rest-api-specs/blob/d4c1ad933ee01abf2f7c3dc0ac4034fc0e6c4317/arm-monitor/2015-04-01/examples/createOrUpdateAutoscaleSetting.json#L14 example

**Review by vishrutshah — CHANGES_REQUESTED** [review]

File: arm-trafficmanager/2017-03-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-03-01/swagger/trafficmanager.json]: Why no name this operation as `Profiles_ListByResourceGroup` ?

File: arm-trafficmanager/2017-05-01/swagger/trafficmanager.json
Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: https://github.com/Azure/azure-sdk-for-net/blob/vs17Dev/src/SDKs/TrafficManager/Management.TrafficManager/Generated/ProfilesOperations.cs#L94
I am referring that Is this code necessary ?

Comment (vladca) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: I understand now, thanks

Comment (vishrutshah) [inline arm-trafficmanager/2017-05-01/swagger/trafficmanager.json]: Based on the offline mail thread, I just want to make sure that we fix this path to have semantically correct swagger and I'll.be happy to merge in the changes. Let me know once you update this. Thanks!

**Comment by vishrutshah** [comment]
@vladca Thanks for fixing the issue. Your changes looks perfect, I am updating CI to make sure things pass

**Comment by vishrutshah** [comment]
Closing and reopening the PR to run the latest npm autorest bootstrapper

**Review by vishrutshah — APPROVED** [review]

**Pull request merged**
